keps/sig-auth/NNNN-20201209-client-exec-proxy/README.md

```diff
@@ -0,0 +1,711 @@
+
+1. Who is responsible for creating the unix domain socket?
+
+   The listening unix domain socket is created by the proxy plugin process.
+
+1. What does the the entire process look like?
+
+   The client attempts a connection to the unix domain socket address (file path) and if it fails, it executes the proxy plugin process, passing the listening address.  The listening unix domain socket is then created by the proxy plugin process and bound to the file path that is passed by the client.  The client then connects to address and upon successful connection, the client sends all requests destined for the kube-apiserver to the proxy plugin.
+
+#### Extending ExecConfig
+
+This proposal extends the ExecConfig API object to include configuration options for the proxy process.
+
+The ExecConfig would be extended to include a local address, which will be passed to the exec plugin.  The exec plugin will create a socket and listen on this address, which can either be a Unix domain socket.
+
+```go
+type ExecConfig struct {
+	// Command to execute.
+	Command string `json:"command"`
+	// Arguments to pass to the command when executing it.
+	// +optional
+	Args []string `json:"args"`
+	// Env defines additional environment variables to expose to the process. These
+	// are unioned with the host's environment, as well as variables client-go uses
+	// to pass argument to the plugin.
+	// +optional
+	Env []ExecEnvVar `json:"env"`
+
+	// Preferred input version of the ExecInfo. The returned ExecCredentials MUST use
+	// the same encoding version as the input.
+	APIVersion string `json:"apiVersion,omitempty"`
+
+	// ...
+
+	// The configuration for an exec proxy.
+	ExecProxy ExecProxyConfig `json:"execProxy"`
+}
+
+```
+
+The ExecProxyConfig includes the address where the proxy is expected to listen.
+```go
+type ExecProxyConfig struct {
+	// The socket path where the proxy should listen.
+	UnixSocketPath string `json:"unixSocketPath"`
+}
+```
+
+Input and output from the executed binary.
+```go
+type ExecProxyRequest struct {
+	metav1.TypeMeta
+
+	// +optional
+	Config ExecProxyConfig `json:"config"`
+
+	// +optional
+	// Whether or not the command is being called from an interactive prompt.  A hardware security module might require user interaction on setup,
+	// so in that case the proxy might want to fail if this is false.
+	Interactive bool `json:"interactive"`
+
+	// +optional
+	// The destination Kubernetes cluster.
+	Cluster *Cluster `json:"cluster"`
+}
+
+type ExecProxyResponse struct {
+	metav1.TypeMeta
+}
+```
+
+Example kubeconfig:
+
+```
+- name: nic@cluster1
+  user:
+    exec:
+      apiVersion: client.authentication.k8s.io/v1alpha1
+      args:
+      - --region=us-west-2
+      command: /usr/local/bin/sigv4proxy
+      execProxy:
+        unixSocketPath: /var/run/kubelet/signer.sock
```

Comment:
This is a non-starter.  I want to be able to download a kubeconfig and have it work on any host that has the exec plugin.

Reply:
I'm going to resolve this because the reference is outdated, but we can continue the discussion about windows and UDS.